src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -199,7 +207,7 @@ public UnfilteredRowIterator computeNext()
                         {
                             UnfilteredRowIterator partition = apply(currentKey);
                             if (partition != null)
-                                return partition;
+                                partitionIterators.add(partition);
```

Comment:
Correctness is IMHO more important than performance and this bug is extremely serious, so I believe in this particular case we should postpone any performance-related work to a separate ticket and merge this dirty fix ASAP. This fix seems to also fix STAR-540 (QueryLifecycleTest). 

Having said that, yes, I agree with you, this increases memory usage and can get bad on very wide partitions.
Unfortunately a hard cutoff would require adding a new setting, and make Cassandra more complex to setup. How would the user establish a good value of the limit? Or maybe we already have a natural limit somewhere, like e.g. per-request fetch size?

Anyway, I feel this could be solved in a different way. There is no reason we need to materialize all the iterators eagerly in advance. The keys coming from the index are already sorted, so we could just create our own UnfilteredRowIterator implementation that would create and concatenate the underlying iterators lazily. This way we get the memory efficiency of the original solution.

Reply:
CC @jacek-lewandowski

Reply:
Ok, I'm good with that as long as we get a ticket raised and are aware of the limitation.